### PR TITLE
PeerGateway connector: Added new methods in the ConnectionProfileDefinition

### PR DESCRIPTION
**Pull request by fraVlaca**
Added changes to ConnectionProfileDefinition with test for new PeerGateway connetor.
The following methods were added:
- getPeersListForOrganization()
- getTlsCertForPeer(peer)
- getGrpcEndPointForPeer(peer)
- getGrpcOptionForPeer(peer)
Furthermore, connector-versions/peer-gateway was added to nyc of caliper-fabric.

### CODE REVIEW

File: packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js [hunk 1]
```diff
@@ -436,4 +436,394 @@ describe('A Connection Profile Definition', async () => {
...
+                discover: false
+            });
+            (() => {
+                connectionProfileDefinition.getGrpcOptionForPeer(peer);
+            }).should.throw(
+                'No peers property can be found in the connection profile provided'
+            );
+        });
+
+        it('should through an error if no grpcOption property for the provided peer is found', async () => {
```
Comment (davidkel) [inline packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js:811]: grpcOptions are optional

**Review by davidkel — CHANGES_REQUESTED** [review]
For general coding a lot of the lines are like this
```
if (!this.connectionProfile.client){
```

The general pattern of the caliper code base is to have a space before the `{`, ie so it looks like
```
if (!this.connectionProfile.client) {
```

File: packages/caliper-fabric/lib/connector-configuration/ConnectionProfileDefinition.js [hunk 2]
```diff
@@ -193,6 +193,128 @@ class ConnectionProfileDefinition {
...
+        return peersInOrg;
+    }
+
+    /**
+     * Return the tls certificate for the specified peer, can only be called when a grps url and respective certs are provided
+     * @param {*} peer the name of the channel
+     * @returns {string} tls certificate
+     */
+    async getTlsCertForPeer(peer){
+        if (!this.connectionProfile.peers){
```
[the remark this answers is not in the document]
Comment (fraVlaca) [reply packages/caliper-fabric/lib/connector-configuration/ConnectionProfileDefinition.js:227]: should I really create a function for a one line check? would the function be a check to throw an error or just return true or false? if it is the second I did not really save any lines of code in the end...

File: packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js [hunk 3]
```diff
@@ -436,4 +436,460 @@ describe('A Connection Profile Definition', async () => {
...
+            );
+        });
+    });
+
+
+    describe('when getting the grpc endpoint for a peer', () => {
+
+        const peer = 'peer0.org1.example.com';
+
+        it('should return the parsed for a grpcs url', async () => {
```
Comment (davidkel) [inline packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js:680]: this description is not quite right, but would suggest `should return the correct endpoint for a grpcs url`

File: packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js [hunk 4]
```diff
@@ -436,4 +436,460 @@ describe('A Connection Profile Definition', async () => {
...
+                }
+            };
+            const connectionProfileDefinition = new ConnectionProfileDefinition(mspId, {
+                loadedConnectionProfile: grpcsConnectionProfile,
+                discover: true
+            });
+            connectionProfileDefinition.getGrpcEndPointForPeer(peer).should.deep.equal('localhost:7051');
+        });
+
+        it('should return the parsed url for a grpc url', async () => {
```
Comment (davidkel) [inline packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js:695]: We can change this description to be similar to above `should return the correct endpoint for a grpc url`

File: packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js [hunk 5]
```diff
@@ -436,4 +436,460 @@ describe('A Connection Profile Definition', async () => {
...
+                discover: false
+            });
+            (() => {
+                connectionProfileDefinition.getGrpcEndPointForPeer(peer);
+            }).should.throw(
+                'http://localhost:7051 is not a valid grpc/grpcs url, make sure to prefix grpc:// or grpcs:// at the beginning of the url'
+            );
+        });
+
+
```
Comment (davidkel) [inline packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js:729]: can remove this blank line

File: packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js [hunk 6]
```diff
@@ -436,4 +436,460 @@ describe('A Connection Profile Definition', async () => {
...
+            }).should.throw(
+                `${peer} provided does not have url property provided in the connection Profile`            );
+        });
+    });
+
+    describe('when getting the grpc options for a peer', () => {
+
+        const peer = 'peer0.org1.example.com';
+
+        it('should return them when provided', async () => {
```
Comment (davidkel) [inline packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js:776]: `should return the defined grpcOptions if present`

File: packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js [hunk 7]
```diff
@@ -436,4 +436,460 @@ describe('A Connection Profile Definition', async () => {
...
+                discover: false
+            });
+            (() => {
+                connectionProfileDefinition.getGrpcOptionForPeer(peer);
+            }).should.throw(
+                `${peer} provided is not present in the connection profile`
+            );
+        });
+
+        it('should throw an error if no peers property is found', async () => {
```
Comment (davidkel) [inline packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js:802]: `is found in the connection profile`

File: packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js [hunk 8]
```diff
@@ -436,4 +436,460 @@ describe('A Connection Profile Definition', async () => {
...
+                discover: false
+            });
+            (() => {
+                connectionProfileDefinition.getGrpcOptionForPeer(peer);
+            }).should.throw(
+                'No peers property can be found in the connection profile provided'
+            );
+        });
+
+        it('should return an empty object if no grpcOption property for the provided peer was found', async () => {
```
Comment (davidkel) [inline packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js:814]: grpcOption --> grpcOptions

File: packages/caliper-fabric/lib/connector-configuration/ConnectionProfileDefinition.js [hunk 9]
```diff
@@ -193,6 +193,135 @@ class ConnectionProfileDefinition {
...
+        const url = peerObj.url.toString();
+        return url.startsWith('grpcs://');
+    }
+
+    /**
+     * Return the gRpc options for the specified peer
+     * @param {*} peer the name of the channel
+     * @returns {[*]} the list of grpc object properties
+     */
+    getGrpcOptionForPeer(peer) {
```
Comment (davidkel) [inline packages/caliper-fabric/lib/connector-configuration/ConnectionProfileDefinition.js:301]: method name should be `getGrpcOptionsForPeer` just missing the `s`

File: packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js [hunk 10]
```diff
@@ -436,4 +436,460 @@ describe('A Connection Profile Definition', async () => {
...
+                }
+            };
+            const connectionProfileDefinition = new ConnectionProfileDefinition(mspId, {
+                loadedConnectionProfile: noUrlConnectionProfile,
+                discover: false
+            });
+            (() => {
+                connectionProfileDefinition.getGrpcEndPointForPeer(peer);
+            }).should.throw(
+                `${peer} provided does not have url property provided in the connection Profile`            );
```
Comment (davidkel) [inline packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js:892]: remove the extra blank spaces

**Review by davidkel — CHANGES_REQUESTED** [review]
Sorry, some more changes

File: packages/caliper-fabric/lib/connector-configuration/ConnectionProfileDefinition.js [hunk 11]
```diff
@@ -193,6 +193,135 @@ class ConnectionProfileDefinition {
...
+     * @param {*} mspid the mspId of the org
+     * @returns {[string]} an array containing the list of all the peers
+     */
+    getPeersListForOrganization(mspid) {
+        if (!this.connectionProfile.organizations) {
+            throw new Error('No organizations property can be found for the connection profile provided');
+        }
+        let peers;
+        for (const org in this.connectionProfile.organizations) {
+            if (this.connectionProfile.organizations[org].mspid === mspid){
```
Comment (davidkel) [inline packages/caliper-fabric/lib/connector-configuration/ConnectionProfileDefinition.js:208]: could we add a space before the `{` please

File: packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js [hunk 12]
```diff
@@ -436,4 +436,460 @@ describe('A Connection Profile Definition', async () => {
...
+            };
+            const connectionProfileDefinition = new ConnectionProfileDefinition(mspId, {
+                loadedConnectionProfile: invalidPemConnectionProfile,
+                discover: false
+            });
+            await connectionProfileDefinition.getTlsCertForPeer(peer).should.be.rejectedWith(
+                `pem provided for ${peer} in the connection profile .tlsCACerts.pem is not valid`
+            );
+        });
+        it('should throw an error if path provided through tlsCAcets.path does not point to a valid cert', async () => {
```
[the remark this answers is not in the document]
Comment (fraVlaca) [reply packages/caliper-fabric/test/connector-configuration/ConnectionProfileDefinition.js:655]: one is for the the .path property not pointing to a valid ceer while the other one is for a not valid cert in the .pem property. Should I revome anyway or is it sendible to have two different ones?

**Review by davidkel — APPROVED** [review]
LGTM

**Pull request merged**